sequence_task_1/student/include/impl.h

```diff
@@ -1,51 +1,86 @@
 #pragma once
+#include <stdexcept>
 #include <string>
 
-struct Song
-{
-    std::string name;
-    explicit Song(std::string song_name)
-        : name {std::move(song_name)}
-    {
-    }
+struct Song {
+  std::string name;
+  explicit Song(std::string song_name) : name{std::move(song_name)} {}
 };
 
-bool operator==(const Song& lhs, const Song& rhs)
-{
-    return lhs.name == rhs.name;
+bool operator==(const Song& lhs, const Song& rhs) {
+  return lhs.name == rhs.name;
 }
 
-template<class Container>
-class StaticPlaylist
-{
-public:
-    /** @todo Member traits */
+template <class Container>
+class StaticPlaylist {
+ public:
+  /** @todo Member traits */
+  using size_type = typename Container::size_type;
+  using value_type = typename Container::value_type;
+  using reference = typename Container::reference;
+  using const_reference = typename Container::const_reference;
+  using iterator = typename Container::iterator;
+  using const_iterator = typename Container::const_iterator;
+  using difference_type = typename Container::difference_type;
+
+  /** @todo Iterators */
+  const_iterator begin() {
+    return m_tracklist.begin();
+  }
 
-    /** @todo Iterators */
+  const_iterator end() {
+    return m_tracklist.end();
+  }
 
-    /** @todo Constructor from any reversible sequence container */
+  /** @todo Constructor from any reversible sequence container */
+  template <class T>
+  StaticPlaylist(const T& track_list) {
+    m_tracklist.assign(track_list.begin(), track_list.end());
+  }
 
-    /** @todo Assignment from any reversible sequence container */
+  /** @todo Assignment from any reversible sequence container */
+  template <class T>
+  StaticPlaylist& operator=(const T& track_list) {
+    m_tracklist.assign(track_list.begin(), track_list.end());
+    return *this;
+  }
 
-    /** @todo Add track from initializer */
-    template<class... Args>
-    const Song& play(Args&&... songData);
+  /** @todo Add track from initializer */
+  template <class... Args>
+  const Song& play(Args&&... songData) {
+    m_tracklist.emplace_back(songData...);
```

Comment:
Use std::forward
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f19-for-forward-parameters-pass-by-tp-and-only-stdforward-the-parameter